### PR TITLE
songbird-server

### PR DESCRIPTION
**Pull request by Oleg11713**

### CODE REVIEW

**Review by VladislavTemriuk — CHANGES_REQUESTED** [review]
Need some changes

File: controllers/gameController.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const { Bird } = require("../models/models");
+
+class GameController {
+  async create(req, res) {
+    let { id, name, species, description, image, audio } = req.body;
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/gameController.js:5]: Replaced let with const

File: controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
+const ApiError = require("../error/ApiError");
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/userController.js:1]: Grouped all imports (first from node_modules, then my own files)

File: controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    if (!email || !password) {
+      return next(ApiError.badRequest("Неверный логин или пароль"));
+    }
+    const candidate = await User.findOne({ where: { email } });
+    if (candidate) {
+      return next(
+        ApiError.badRequest("Пользователь с таким email уже существует")
+      );
+    }
+    const hashPassword = await bcrypt.hash(password, 5);
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/userController.js:26]: Made a constant SALT and renamed the "hashPassword" to "passwordHash" for better understanding

File: controllers/userController.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    const token = generateJwt(
+      user.id,
+      user.displayName,
+      user.email,
+      user.totalScoreForAllGames
+    );
+    return res.json({ token });
+  }
+
+  async check(req, res, next) {
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/userController.js:72]: Initially, I wanted to make an additional token verification function, but later I decided that I didn’t need it, and forgot to remove it
I also didn't need the function getAll in userController
Removed unused features

File: db.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+module.exports = new Sequelize(
+  process.env.DB_NAME,
+  process.env.DB_USER,
+  process.env.DB_PASSWORD,
+  {
+    dialect: "postgres",
+    host: process.env.DB_HOST,
+    port: process.env.DB_PORT,
+  }
+);
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply db.js:12]: Done

File: models/models.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+});
+
+const Bird = sequelize.define("bird", {
+  id: { type: DataTypes.INTEGER, primaryKey: true },
+  name: { type: DataTypes.STRING, allowNull: false },
+  species: { type: DataTypes.STRING, allowNull: false },
+  description: { type: DataTypes.STRING, allowNull: false },
+  image: { type: DataTypes.STRING, allowNull: false },
+  audio: { type: DataTypes.STRING, allowNull: false },
+});
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply models/models.js:19]: Separated models into files userModel and birdModel

File: routes/index.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+const Router = require("express");
+const router = new Router();
+const gameRouter = require("./gameRouter");
+const userRouter = require("./userRouter");
+
+router.use("/signIn", userRouter);
+router.use("/", gameRouter);
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply routes/index.js:7]: Done

File: .env.example [hunk 8]
```diff
@@ -0,0 +1,7 @@
+PORT
+DB_NAME
+DB_USER
+DB_PASSWORD
+DB_HOST
+DB_PORT
+SECRET_KEY
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply .env.example:7]: Renamed to ACCESS_TOKEN_KEY

File: controllers/userController.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/userController.js:72]: The function checks the token, if it does not exist, the middlewareAuth gives an error 401, I forgot to write the function in the client, I have already added it
Also removed unused function getAll in userController

File: controllers/userController.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  return jwt.sign(
+    { id, displayName, email, totalScoreForAllGames },
+    process.env.SECRET_KEY,
+    { expiresIn: "24h" }
+  );
+};
+
+class UserController {
+  async registration(req, res, next) {
+    const { displayName, email, password, totalScoreForAllGames } = req.body;
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/userController.js:16]: Done

File: middleware/authMiddleware.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+module.exports = function (req, res, next) {
+  if (req.method === "OPTION") {
+    next();
+  }
+  try {
+    const token = req.headers.authorization.split(" ")[1];
+    if (!token) {
+      return res.status(401).json({ message: "Пользователь не авторизован" });
+    }
+    req.user = jwt.verify(token, process.env.ACCESS_TOKEN_KEY);
```
Comment (VladislavTemriuk) [inline middleware/authMiddleware.js:12]: you should check if token didn't pass a verification

File: routes/gameRouter.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+const Router = require("express");
+const router = new Router();
+
+const gameController = require("../controllers/gameController");
+const userController = require("../controllers/userController");
+
+router.post("/addBird", gameController.create);
+router.patch("/", userController.updateTotalScore);
```
Comment (VladislavTemriuk) [inline routes/gameRouter.js:8]: maybe also add path for this route? for example `/updateScore`

Comment (VladislavTemriuk) [reply routes/gameRouter.js:8]: also you should add middleware for this routes, because without it everyone can call it and change data

**Review by VladislavTemriuk — CHANGES_REQUESTED** [review]
Good job

but need some changes

File: controllers/gameController.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    const bird = await Bird.create({
+      id,
+      name,
+      species,
+      description,
+      image,
+      audio,
+    });
+    return res.json(bird);
+  }
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/gameController.js:15]: Initially, I created a function to check whether requests are working correctly, now it is really not needed (removed)

File: controllers/gameController.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+const { Bird } = require("../models/birdModel");
+
+class GameController {
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/gameController.js:3]: Looked again, really wrong name (renamed)

File: controllers/userController.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+  async login(req, res, next) {
+    const { email, password } = req.body;
+    const user = await User.findOne({ where: { email } });
+    if (!user) {
+      return next(
+        ApiError.badRequest("Пользователя с таким email не существует")
+      );
+    }
+    const comparePassword = await bcrypt.compareSync(password, user.password);
```
[the remark this answers is not in the document]
Comment (Oleg11713) [reply controllers/userController.js:63]: Corrected

File: middleware/authMiddleware.js (hunk 10 above)
Comment (Oleg11713) [reply middleware/authMiddleware.js:12]: Done

File: routes/gameRouter.js (hunk 11 above)
Comment (Oleg11713) [reply routes/gameRouter.js:8]: Added path for route ('/updateScore')

Comment (Oleg11713) [reply routes/gameRouter.js:8]: Added middleware for this route to check who is trying to change

File: routes/userRouter.js [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+const authMiddleware = require("../middleware/authMiddleware");
+const checkForUpdateMiddleware = require("../middleware/checkForUpdateMiddleware");
+
+const router = new Router();
+
+router.post("/registration", userController.registration);
+router.post("/login", userController.login);
+router.patch(
+  "/updateScore",
+  checkForUpdateMiddleware,
```
Comment (VladislavTemriuk) [inline routes/userRouter.js:13]: i'm not sure that you check for update, you check if user is authorised. So better to name it authMiddleware

Comment (VladislavTemriuk) [reply routes/userRouter.js:13]: and you have almost the same middleware's, just with different error messages and statuses, so if you need them both, better to pass arguments to not duplicate the code

**Review by VladislavTemriuk — CHANGES_REQUESTED** [review]
Good job👍

**Pull request closed without merging**